File: scripts/day6-part1-simple.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <chrono>
#include <sstream>

using namespace std;

struct Problem {
    vector<long long> numbers;
    char op;
};

long long solvePart1(const string& input) {
    stringstream ss(input);
    vector<string> lines;
    string line;
    
    while (getline(ss, line)) {
        lines.push_back(line);
    }
    
    if (lines.empty()) return 0;
    
    int width = lines[0].length();
    int height = lines.size();
    
    vector<Problem> problems;
    int col = 0;
    
    while (col < width) {
        // Check if this column starts a problem
        bool hasNonSpace = false;
        for (int row = 0; row < height; row++) {
            if (col < lines[row].length() && lines[row][col] != ' ') {
                hasNonSpace = true;
                break;
            }
        }
        
        if (!hasNonSpace) {
            col++;
            continue;
        }
        
        // Find the end of this problem
        int endCol = col;
        while (endCol < width) {
            bool isAllSpace = true;
            for (int row = 0; row < height; row++) {
                if (endCol < lines[row].length() && lines[row][endCol] != ' ') {
                    isAllSpace = false;
                    break;
                }
            }
            if (isAllSpace) break;
            endCol++;
        }
        
        // Extract this problem
        vector<long long> numbers;
        char op = '+';
        
        for (int row = 0; row < height; row++) {
            string cell = lines[row].substr(col, endCol - col);
            // Trim
            size_t start = cell.find_first_not_of(' ');
            size_t end = cell.find_last_not_of(' ');
            if (start != string::npos) {
                cell = cell.substr(start, end - start + 1);
                
                if (cell == "+" || cell == "*") {
                    op = cell[0];
                } else if (!cell.empty()) {
                    numbers.push_back(stoll(cell));
                }
            }
        }
        
        problems.push_back({numbers, op});
        col = endCol;
    }
    
    // Calculate answers
    long long grandTotal = 0;
    
    for (const auto& problem : problems) {
        long long result;
        if (problem.op == '+') {
            result = 0;
            for (long long num : problem.numbers) {
                result += num;
            }
        } else {
            result = 1;
            for (long long num : problem.numbers) {
                result *= num;
            }
        }
        grandTotal += result;
    }
    
    return grandTotal;
}

int main() {
    // Read input
    ifstream file("../inputs/day-6-input.txt");
    stringstream buffer;
    buffer << file.rdbuf();
    string input = buffer.str();
    file.close();
    
    // Remove trailing newline
    if (!input.empty() && input.back() == '\n') {
        input.pop_back();
    }
    if (!input.empty() && input.back() == '\r') {
        input.pop_back();
    }
    
    // Benchmark
    const int iterations = 1;
    auto start = chrono::high_resolution_clock::now();
    
    long long result;
    for (int i = 0; i < iterations; i++) {
        result = solvePart1(input);
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(end - start);
    double avgMicros = duration.count() / (double)iterations / 1000.0;
    
    cout << "C++ (Day 6 Part 1 - Simple)" << endl;
    cout << "Answer: " << result << endl;
    cout << "Average time: " << avgMicros << " microseconds (" << iterations << " iterations)" << endl;
    cout << "Total time: " << (duration.count() / 1000000.0) << " ms" << endl;
    
    return 0;
}

File: scripts/day6-part1-simple.js
const fs = require("fs");

function solvePart1(input) {
  const lines = input.split("\n");
  const width = lines[0].length;
  const height = lines.length;

  // Find problem boundaries
  const problems = [];
  let col = 0;

  while (col < width) {
    // Check if this column starts a problem
    let hasNonSpace = false;
    for (let row = 0; row < height; row++) {
      if (lines[row][col] && lines[row][col] !== " ") {
        hasNonSpace = true;
        break;
      }
    }

    if (!hasNonSpace) {
      col++;
      continue;
    }

    // Find the end of this problem
    let endCol = col;
    while (endCol < width) {
      let isAllSpace = true;
      for (let row = 0; row < height; row++) {
        if (lines[row][endCol] && lines[row][endCol] !== " ") {
          isAllSpace = false;
          break;
        }
      }
      if (isAllSpace) break;
      endCol++;
    }

    // Extract this problem
    const numbers = [];
    let operator = null;

    for (let row = 0; row < height; row++) {
      const cell = lines[row].substring(col, endCol).trim();
      if (cell) {
        if (cell === "+" || cell === "*") {
          operator = cell;
        } else {
          numbers.push(parseInt(cell));
        }
      }
    }

    problems.push({ numbers, operator });
    col = endCol;
  }

  // Calculate answers
  let grandTotal = 0;

  for (const problem of problems) {
    let result;
    if (problem.operator === "+") {
      result = problem.numbers.reduce((a, b) => a + b, 0);
    } else {
      result = problem.numbers.reduce((a, b) => a * b, 1);
    }
    grandTotal += result;
  }

  return grandTotal;
}

const input = fs.readFileSync("../inputs/day-6-input.txt", "utf-8").trim();

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart1(input);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 6 Part 1 - Simple)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);
